src/viem/prepareTransactions.ts

```diff
@@ -0,0 +1,155 @@
+import { TokenBalance, TokenBalanceWithAddress } from 'src/tokens/slice'
+import { TransactionRequestCIP42 } from 'node_modules/viem/_types/chains/celo/types'
+import BigNumber from 'bignumber.js'
+import { Address, EstimateGasExecutionError } from 'viem'
+import { estimateFeesPerGas } from 'src/viem/estimateFeesPerGas'
+import { publicClient } from 'src/viem/index'
+import { STATIC_GAS_PADDING } from 'src/config'
+import Logger from 'src/utils/Logger'
+
+const TAG = 'viem/prepareTransactions'
+
+interface PreparedTransactionsPossible {
+  type: 'possible'
+  transactions: TransactionRequestCIP42[]
+}
+
+export interface PreparedTransactionsNeedDecreaseSpendAmountForGas {
+  type: 'need-decrease-spend-amount-for-gas'
+  maxGasCost: BigNumber
+  feeCurrency: TokenBalance
+  decreasedSpendAmount: BigNumber
+}
+
+export interface PreparedTransactionsNotEnoughBalanceForGas {
+  type: 'not-enough-balance-for-gas'
+  feeCurrencies: TokenBalance[]
+}
+
+export type PreparedTransactionsResult =
+  | PreparedTransactionsPossible
+  | PreparedTransactionsNeedDecreaseSpendAmountForGas
+  | PreparedTransactionsNotEnoughBalanceForGas
+
+function getMaxGasCost(txs: TransactionRequestCIP42[]): BigNumber {
+  let maxGasCost = BigInt(0)
+  for (const tx of txs) {
+    if (!tx.gas || !tx.maxFeePerGas) {
+      throw new Error('Missing gas or maxFeePerGas')
+    }
+    maxGasCost += BigInt(tx.gas) * BigInt(tx.maxFeePerGas)
+  }
+  return new BigNumber(maxGasCost.toString())
+}
+
+function allTruthy<T>(arr: (T | undefined | null)[]): arr is T[] {
+  return arr.every((x) => !!x)
+}
+
+export async function prepareTransactions({
+  feeCurrencies,
+  spendToken,
+  spendTokenAmount,
+  decreasedAmountGasCostMultiplier,
+  baseTransactions,
+}: {
+  feeCurrencies: TokenBalance[]
+  spendToken: TokenBalanceWithAddress
+  spendTokenAmount: BigNumber
+  decreasedAmountGasCostMultiplier: number
+  baseTransactions: (TransactionRequestCIP42 & { gas?: bigint })[]
+}): Promise<PreparedTransactionsResult> {
+  const maxGasCosts: Array<{ feeCurrency: TokenBalance; maxGasCostInDecimal: BigNumber }> = []
+  for (const feeCurrency of feeCurrencies) {
+    if (feeCurrency.balance.isLessThanOrEqualTo(0)) {
+      // No balance, try next fee currency
+      continue
+    }
+    const feeCurrencyAddress = !feeCurrency.isNative ? (feeCurrency.address as Address) : undefined
+    const transactions: (TransactionRequestCIP42 | null)[] = await Promise.all(
+      baseTransactions.map(async (baseTransaction) => {
+        const { maxFeePerGas, maxPriorityFeePerGas } = await estimateFeesPerGas(
+          publicClient.celo,
+          feeCurrencyAddress
+        )
+        const transaction = {
+          ...baseTransaction,
+          maxFeePerGas,
+          maxPriorityFeePerGas,
+          feeCurrency: feeCurrencyAddress,
+        }
+        if (baseTransaction.gas) {
```

Comment:
this is the main difference between the generalized `prepareTransactions` and its previous form that was specific to swaps. It lets the consumer tell this helper whether gas should be estimated with viem, or a pre-calculated value should be used (as in the case of swaps, where you want to check with viem for "approve" but use a precalculated string for "swap")

Reply:
Yeah this is great. One thing is I think we shouldn't process the base transactions in parallel.
And only estimate the first one if gas is not set. 
As they can be dependant transactions, like for swaps.
I had started doing this in my local branch for swaps just yesterday :D

Reply:
This is what I was proposing about this specific point: https://github.com/valora-inc/wallet/pull/4368/commits/327adddbf846b41be9e8ffc8042ef028fa0484c3#diff-c9b55f4aa0b75764dddb8dac1eb9aa7df00d6195311c998a0df0a95bfd726eacR116-R162